### PR TITLE
Application refactor migration

### PR DESCRIPTION
**Pull request by tim-dot-org**
Just the migration for related PR: https://github.com/code-dot-org/code-dot-org/pull/42334
Original error:
```ActiveRecord::SubclassNotFound: The single-table inheritance mechanism failed to locate the subclass: 'Pd::Application::Teacher2122Application'. This error is raised because the column 'type' is reserved for storing the class in case of inheritance. ```

### CODE REVIEW

File: dashboard/app/models/code_review_comment.rb [hunk 1]
```diff
@@ -8,7 +8,7 @@
 #  script_id        :integer
 #  level_id         :integer
 #  commenter_id     :integer          not null
-#  comment          :text(16777215)
+#  comment          :text(65535)
```
Comment (bethanyaconnor) [inline dashboard/app/models/code_review_comment.rb]: Do you mind doing these as a separate PR? I don't see diffs locally when I migrate on staging but I do see diffs to these files on your branch. I don't know why these are being annotated differently in different environments, but it's probably good to work that out

**Review by bethanyaconnor — APPROVED** [review]
Looks good to me! Thanks for splitting this out!

File: dashboard/app/models/code_review_comment.rb (hunk 1 above)
Comment (breville) [reply dashboard/app/models/code_review_comment.rb]: I think it's related to UTF8 encoding.  @jamescodeorg had some expertise in this area...

Comment (jamescodeorg) [reply dashboard/app/models/code_review_comment.rb]: Hmmm, I see this locally too and never dug into this.  I found this issue that's probably relevant: https://github.com/ctran/annotate_models/issues/278

Comment (jamescodeorg) [reply dashboard/app/models/code_review_comment.rb]: Looks like the annotate gem is [getting this value](https://github.com/ctran/annotate_models/blob/develop/lib/annotate/annotate_models.rb#L881) from rails.

```
Locally:
[development] dashboard > CodeReviewComment.columns[6]
=> #<ActiveRecord::ConnectionAdapters::MySQL::Column:0x00007f94f06c2aa0 @name="comment", @table_name="code_review_comments", @sql_type_metadata=#<ActiveRecord::ConnectionAdapters::SqlTypeMetadata:0x00007f94f06c2c80 @sql_type="text", @type=:text, @limit=65535, @precision=nil, @scale=nil>, @null=true, @default=nil, @default_function=nil, @collation="utf8mb4_bin", @comment=nil>

Staging:
[staging] dashboard > CodeReviewComment.columns[6]
=> #<ActiveRecord::ConnectionAdapters::MySQL::Column:0x00005641279b5320 @name="comment", @table_name="code_review_comments", @sql_type_metadata=#<ActiveRecord::ConnectionAdapters::SqlTypeMetadata:0x00005641279b4fd8 @sql_type="mediumtext", @type=:text, @limit=16777215, @precision=nil, @scale=nil>, @null=true, @default=nil, @default_function=nil, @collation="utf8mb4_bin", @comment=nil>
```

Interestingly, the comment column has a different type locally vs. on staging.  A quick google search didn't turn up why this might be the case. 

(Ok, I'm going to stop procrastinating now.  :-)

File: dashboard/db/migrate/20210907200327_convert_application_classes.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+  # Application subclasses were consolidated from year specific classes to a single application class
+  # Rails stores a db field 'type' which indicates which ruby class should be associated with the db entry,
+  # so these types need to be updated to match the classes that exist.
+  def change
+    reversible do |dir|
+      dir.up do
+        # convert Teacher1819Application to TeacherApplication
+        ActiveRecord::Base.transaction do
+          Pd::Application::ApplicationBase.
```
Comment (bethanyaconnor) [inline dashboard/db/migrate/20210907200327_convert_application_classes.rb:23]: Sorry for commenting late, but are we worried about this migration taking a long time? I figured the table isn't super large but I'm not sure

Comment (tim-dot-org) [reply dashboard/db/migrate/20210907200327_convert_application_classes.rb:23]: there are 42k entries in pd_applications, so I'm not super concerned about this migration, especially the up direction which uses update_all which i think is more performant

Comment (bethanyaconnor) [reply dashboard/db/migrate/20210907200327_convert_application_classes.rb:23]: Yeah that seems fine to me! I did a migration with an `update_all` recently that ran into problems, but that table had a few million entries.

**Comment by tim-dot-org** [comment]
Tested against the production database, and it didn't work for deleted entries! so I fixed that, and made it more resilient in the down direction.

File: dashboard/db/migrate/20210907200327_convert_application_classes.rb (hunk 2 above)
Comment (tim-dot-org) [reply dashboard/db/migrate/20210907200327_convert_application_classes.rb:23]: migration took ~10 seconds on a production clone

**Comment by bencodeorg** [comment]
> Tested against the production database, and it didn't work for deleted entries! so I fixed that, and made it more resilient in the down direction.

Nice! And were you able to check out core application functionality (eg, reviewing applications, submitting applications, etc.) on the adhoc?

**Comment by tim-dot-org** [comment]
> 
> 
> > Tested against the production database, and it didn't work for deleted entries! so I fixed that, and made it more resilient in the down direction.
> 
> Nice! And were you able to check out core application functionality (eg, reviewing applications, submitting applications, etc.) on the adhoc?

Yep, tested that a bit more with the prod data (i didn't actually submit an application, because i think being tied to a production clone it would have actually emailed the regional partner) but loading the dashboard, viewing individual applications, and going through filling out the application worked

**Review by bencodeorg — APPROVED** [review]
🎉

**Review by Sam-Hutchins — APPROVED** [review]
Nothing to add here :)
